lib/senders/email.js

```diff
@@ -253,97 +267,227 @@ module.exports = function (log, config) {
+  //
+  // It will attempt to read live config data from Redis and live config takes
+  // precedence over local static config. If no config is found at all, email
+  // will be routed locally via the auth server.
+  //
+  // Live config looks like this (every property is optional):
+  //
+  // {
+  //   sendgrid: {
+  //     percentage: 100,
+  //     regex: "^.+@example\.com$"
+  //   },
+  //   ses: {
+  //     percentage: 10,
+  //     regex: ".*",
+  //   }
+  // }
+  //
+  // Where a percentage and a regex are both present, an email address must
+  // satisfy both criteria to count as a match. Where an email address matches
+  // both sendgrid and ses, sendgrid wins.
+  //
+  // @param {Object} message
+  //
+  // @returns {Promise} Resolves to an array of `Service` objects.
+  //
+  // @typedef {Object} Service
+  //
+  // @property {Object} mailer           The object on which to invoke the `sendMail`
+  //                                     method.
+  //
+  // @property {String[]} emailAddresses The array of email addresses to send to.
+  //                                     The address at index 0 will be used as the
+  //                                     `to` address and any remaining addresses
+  //                                     will be included as `cc` addresses.
+  //
+  // @property {String} emailService     The name of the email service for metrics.
+  //
+  // @property {String} emailSender      The name of the underlying email sender,
+  //                                     used for both metrics and sent as the
+  //                                     `provider` param in external requests.
+  Mailer.prototype.selectEmailServices = function (message) {
+    const mailers = {
+      internal: this.mailer,
+      external: this.emailService
+    }
+    const indices = new Map()
+    const emailAddresses = [ message.email ]
+    if (Array.isArray(message.ccEmails)) {
+      emailAddresses.push(...message.ccEmails)
+    }
+
+    return redis.get('config')
+      .catch(() => {})
+      .then(liveConfig => {
+        return emailAddresses.reduce((services, emailAddress) => {
+          if (liveConfig) {
+            const { sendgrid, ses } = liveConfig
+
+            if (sendgrid && isLiveConfigMatch(sendgrid, emailAddress)) {
+              return upsertServicesArrayExternalSendgrid(services, emailAddress)
             }
-          )
-          return d.reject(err)
-        }
-        log.info(
-          {
-            op: 'mailer.send.1',
-            status: status && status.message,
-            id: status && status.messageId,
-            to: emailConfig && emailConfig.to
+
+            if (ses && isLiveConfigMatch(ses, emailAddress)) {
+              return upsertServicesArrayExternalSes(services, emailAddress)
+            }
+          }
+
+          if (config.emailService.forcedEmailAddresses.test(message.email)) {
+            return upsertServicesArrayExternalSes(services, emailAddress)
```

Comment:
Ah, so if no liveConfig, `emailService.forcedEmailAddresses` always go to SES, but via the emailService instead of internal?